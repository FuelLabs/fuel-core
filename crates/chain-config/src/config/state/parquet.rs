pub mod decode;
pub mod encode;

#[cfg(test)]
mod tests {
    use bytes::Bytes;
    use itertools::Itertools;
    use parquet::{
        basic::ZstdLevel,
        file::reader::{
            ChunkReader,
            Length,
        },
    };
    use rand::{
        rngs::StdRng,
        Rng,
        SeedableRng,
    };
    use std::{
        io::{
            Cursor,
            Read,
            Seek,
            SeekFrom,
        },
        iter::repeat_with,
        sync::{
            atomic::AtomicU64,
            Arc,
            Mutex,
        },
    };

    use super::{
        decode::*,
        encode::*,
    };

    #[derive(Clone)]
    struct TrackingReader {
        source: Arc<Mutex<Cursor<Vec<u8>>>>,
        read_bytes: Arc<AtomicU64>,
    }

    impl TrackingReader {
        fn new(source: Vec<u8>) -> Self {
            Self {
                source: Arc::new(Mutex::new(Cursor::new(source))),
                read_bytes: Default::default(),
            }
        }
    }

    impl Seek for TrackingReader {
        fn seek(&mut self, pos: SeekFrom) -> std::io::Result<u64> {
            self.source.lock().unwrap().seek(pos)
        }
    }

    impl Read for TrackingReader {
        fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
            let amount = self.source.lock().unwrap().read(buf)?;
            self.read_bytes
                .fetch_add(amount as u64, std::sync::atomic::Ordering::SeqCst);
            Ok(amount)
        }
    }

    impl Length for TrackingReader {
        fn len(&self) -> u64 {
            let mut source = self.source.lock().unwrap();

            let current = source.stream_position().unwrap();

            source.seek(SeekFrom::Start(0)).unwrap();
            let len = source.seek(SeekFrom::End(0)).unwrap();

            source.seek(SeekFrom::Start(current)).unwrap();
            len
        }
    }

    impl ChunkReader for TrackingReader {
        type T = TrackingReader;

        fn get_read(&self, start: u64) -> parquet::errors::Result<Self::T> {
            let mut new_reader = self.clone();
            new_reader.seek(SeekFrom::Start(start))?;

            Ok(new_reader)
        }

        fn get_bytes(&self, start: u64, length: usize) -> parquet::errors::Result<Bytes> {
            let mut buf = Vec::with_capacity(length);
            let reader = self.get_read(start)?;
            reader.take(length as u64).read_to_end(&mut buf)?;

            Ok(Bytes::from(buf))
        }
    }

    #[test]
    fn can_skip_groups_without_reading_whole_file() {
        // given
        let mut buffer = vec![];
        let mut encoder = Encoder::new(
            &mut buffer,
            parquet::basic::Compression::ZSTD(ZstdLevel::try_new(1).unwrap()),
        )
        .unwrap();
        let mut rng = StdRng::seed_from_u64(0);

        let big_group = repeat_with(|| rng.gen::<[u8; 32]>().to_vec())
            .take(1000)
            .collect_vec();
        encoder.write(big_group).unwrap();

        let small_group = vec![rng.gen::<[u8; 32]>().to_vec()];
        encoder.write(small_group).unwrap();
        encoder.close().unwrap();
        let total_size = buffer.len();

        let bytes = TrackingReader::new(buffer);
        let bytes_read = bytes.read_bytes.clone();

        let mut decoder = Decoder::new(bytes).unwrap();

        // when
        let _: Vec<_> = decoder.nth(1).unwrap().unwrap();

        // then
        let actually_read = bytes_read.load(std::sync::atomic::Ordering::SeqCst);

        assert_eq!(total_size, 36930);
        assert_eq!(actually_read, 509);
    }
}
