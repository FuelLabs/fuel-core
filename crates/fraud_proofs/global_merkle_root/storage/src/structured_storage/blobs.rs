use crate::{
    column::TableColumn,
    merkle::MerkleizedTableColumn,
};
use fuel_core_storage::tables::BlobData;

impl MerkleizedTableColumn for BlobData {
    fn table_column() -> TableColumn {
        TableColumn::Blobs
    }
}
