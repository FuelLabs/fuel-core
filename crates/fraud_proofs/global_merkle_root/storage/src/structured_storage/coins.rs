use crate::{
    column::TableColumn,
    merkle::MerkleizedTableColumn,
};
use fuel_core_storage::tables::Coins;

impl MerkleizedTableColumn for Coins {
    fn table_column() -> TableColumn {
        TableColumn::Coins
    }
}
