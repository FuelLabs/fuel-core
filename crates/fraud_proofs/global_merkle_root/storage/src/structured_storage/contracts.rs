use crate::{
    column::TableColumns,
    merkle::MerklizedTableColumn,
};
use fuel_core_storage::tables::{
    ContractsLatestUtxo,
    ContractsRawCode,
};

impl MerklizedTableColumn for ContractsRawCode {
    fn table_column() -> TableColumns {
        TableColumns::ContractsRawCode
    }
}

impl MerklizedTableColumn for ContractsLatestUtxo {
    fn table_column() -> TableColumns {
        TableColumns::ContractsLatestUtxo
    }
}
