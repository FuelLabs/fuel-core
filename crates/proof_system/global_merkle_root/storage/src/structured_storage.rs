mod blobs;
mod coins;
mod contracts;
mod messages;
mod transactions;
mod upgrades;
