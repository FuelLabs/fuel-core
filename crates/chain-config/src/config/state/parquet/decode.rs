use anyhow::{
    anyhow,
    Context,
};
use parquet::{
    data_type::AsBytes,
    file::{
        reader::{
            ChunkReader,
            FileReader,
        },
        serialized_reader::SerializedFileReader,
    },
    record::RowAccessor,
};

pub struct Decoder<R: ChunkReader> {
    data_source: SerializedFileReader<R>,
    group_index: usize,
}

impl<R> Decoder<R>
where
    R: ChunkReader + 'static,
{
    pub fn num_groups(&self) -> usize {
        self.data_source.num_row_groups()
    }

    fn current_group(&self) -> anyhow::Result<Vec<Vec<u8>>> {
        let data = self
            .data_source
            .get_row_group(self.group_index)?
            .get_row_iter(None)?
            .map(|result| {
                result.map_err(|e| anyhow!(e)).and_then(|row| {
                    const FIELD_IDX: usize = 0;
                    Ok(row
                        .get_bytes(FIELD_IDX)
                        .context("While decoding postcard bytes")?
                        .as_bytes()
                        .to_vec())
                })
            })
            .collect::<Result<Vec<_>, _>>()?;

        Ok(data)
    }
}

impl<R> Iterator for Decoder<R>
where
    R: ChunkReader + 'static,
{
    type Item = anyhow::Result<Vec<Vec<u8>>>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.group_index >= self.data_source.metadata().num_row_groups() {
            return None;
        }

        let group = self.current_group();
        self.group_index = self.group_index.saturating_add(1);

        Some(group)
    }

    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.group_index = self.group_index.saturating_add(n);
        self.next()
    }
}

impl<R: ChunkReader + 'static> Decoder<R> {
    pub fn new(reader: R) -> anyhow::Result<Self> {
        Ok(Self {
            data_source: SerializedFileReader::new(reader)?,
            group_index: 0,
        })
    }
}
