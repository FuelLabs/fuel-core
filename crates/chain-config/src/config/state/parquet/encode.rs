use std::{
    io::Write,
    marker::PhantomData,
    sync::Arc,
};

use itertools::Itertools;
use parquet::{
    basic::{
        Compression,
        Repetition,
    },
    file::{
        properties::WriterProperties,
        writer::SerializedFileWriter,
    },
};

use parquet::{
    data_type::ByteArrayType,
    schema::types::Type,
};
use postcard::ser_flavors::{
    AllocVec,
    Flavor,
};

pub type PostcardEncoder<T> = Encoder<std::fs::File, T, PostcardEncode>;

pub struct Encoder<W: Write + Send, T, E> {
    writer: SerializedFileWriter<W>,
    _type: PhantomData<T>,
    _encoding: PhantomData<E>,
}

impl<W: Write + Send, T, E> Encoder<W, T, E>
where
    E: Encode<T>,
{
    pub fn new(writer: W, compression: Compression) -> anyhow::Result<Self> {
        let writer = SerializedFileWriter::new(
            writer,
            Arc::new(Self::single_element_schema()),
            Arc::new(
                WriterProperties::builder()
                    .set_compression(compression)
                    .build(),
            ),
        )?;

        Ok(Self {
            writer,
            _type: PhantomData,
            _encoding: PhantomData,
        })
    }

    fn single_element_schema() -> Type {
        let data =
            Type::primitive_type_builder("data", ::parquet::basic::Type::BYTE_ARRAY)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .expect("This is a valid schema");

        Type::group_type_builder("unimportant")
            .with_fields(vec![Arc::new(data)])
            .build()
            .expect("This is a valid schema")
    }
}

impl<W, T, E> Encoder<W, T, E>
where
    W: Write + Send,
    E: Encode<T>,
{
    pub fn write(&mut self, elements: Vec<T>) -> anyhow::Result<()> {
        let mut group = self.writer.next_row_group()?;
        let mut column = group
            .next_column()?
            .ok_or_else(|| anyhow::anyhow!("Missing column. Check the schema!"))?;

        let values: Vec<_> = elements
            .into_iter()
            .map(|el| E::encode(&el))
            .map_ok(Into::into)
            .try_collect()?;
        column
            .typed::<ByteArrayType>()
            .write_batch(&values, None, None)?;

        column.close()?;
        group.close()?;
        Ok(())
    }

    pub fn close(self) -> anyhow::Result<()> {
        self.writer.close()?;
        Ok(())
    }
}

pub trait Encode<T> {
    fn encode(data: &T) -> anyhow::Result<Vec<u8>>;
}

pub struct PostcardEncode;

impl<T> Encode<T> for PostcardEncode
where
    T: serde::Serialize,
{
    fn encode(data: &T) -> anyhow::Result<Vec<u8>> {
        let mut serializer = postcard::Serializer {
            output: AllocVec::new(),
        };
        data.serialize(&mut serializer)?;
        Ok(serializer.output.finalize()?)
    }
}
