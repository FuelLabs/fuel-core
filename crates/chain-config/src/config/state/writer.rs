use crate::{
    config::table_entry::TableEntry,
    AddTable,
    ChainConfig,
    LastBlockConfig,
    SnapshotMetadata,
    StateConfigBuilder,
    TableEncoding,
};
use fuel_core_storage::structured_storage::TableWithBlueprint;
use std::path::PathBuf;

#[cfg(feature = "parquet")]
use super::parquet;

enum EncoderType {
    Json {
        builder: StateConfigBuilder,
    },
    #[cfg(feature = "parquet")]
    Parquet {
        compression: ZstdCompressionLevel,
        table_encoders: TableEncoders,
    },
}

pub struct SnapshotWriter {
    dir: PathBuf,
    encoder: EncoderType,
}

#[allow(dead_code)]
#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    serde::Serialize,
    serde::Deserialize,
)]
#[cfg(feature = "parquet")]
#[cfg_attr(test, derive(strum::EnumIter))]
pub enum ZstdCompressionLevel {
    Uncompressed,
    Level1,
    Level2,
    Level3,
    Level4,
    Level5,
    Level6,
    Level7,
    Level8,
    Level9,
    Level10,
    Level11,
    Level12,
    Level13,
    Level14,
    Level15,
    Level16,
    Level17,
    Level18,
    Level19,
    Level20,
    Level21,
    Max,
}

#[cfg(feature = "parquet")]
impl TryFrom<u8> for ZstdCompressionLevel {
    type Error = anyhow::Error;
    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Uncompressed),
            1 => Ok(Self::Level1),
            2 => Ok(Self::Level2),
            3 => Ok(Self::Level3),
            4 => Ok(Self::Level4),
            5 => Ok(Self::Level5),
            6 => Ok(Self::Level6),
            7 => Ok(Self::Level7),
            8 => Ok(Self::Level8),
            9 => Ok(Self::Level9),
            10 => Ok(Self::Level10),
            11 => Ok(Self::Level11),
            12 => Ok(Self::Level12),
            13 => Ok(Self::Level13),
            14 => Ok(Self::Level14),
            15 => Ok(Self::Level15),
            16 => Ok(Self::Level16),
            17 => Ok(Self::Level17),
            18 => Ok(Self::Level18),
            19 => Ok(Self::Level19),
            20 => Ok(Self::Level20),
            21 => Ok(Self::Level21),
            22 => Ok(Self::Max),
            _ => {
                anyhow::bail!("Compression level {value} outside of allowed range 0..=22")
            }
        }
    }
}

#[cfg(feature = "parquet")]
impl From<ZstdCompressionLevel> for u8 {
    fn from(value: ZstdCompressionLevel) -> Self {
        match value {
            ZstdCompressionLevel::Uncompressed => 0,
            ZstdCompressionLevel::Level1 => 1,
            ZstdCompressionLevel::Level2 => 2,
            ZstdCompressionLevel::Level3 => 3,
            ZstdCompressionLevel::Level4 => 4,
            ZstdCompressionLevel::Level5 => 5,
            ZstdCompressionLevel::Level6 => 6,
            ZstdCompressionLevel::Level7 => 7,
            ZstdCompressionLevel::Level8 => 8,
            ZstdCompressionLevel::Level9 => 9,
            ZstdCompressionLevel::Level10 => 10,
            ZstdCompressionLevel::Level11 => 11,
            ZstdCompressionLevel::Level12 => 12,
            ZstdCompressionLevel::Level13 => 13,
            ZstdCompressionLevel::Level14 => 14,
            ZstdCompressionLevel::Level15 => 15,
            ZstdCompressionLevel::Level16 => 16,
            ZstdCompressionLevel::Level17 => 17,
            ZstdCompressionLevel::Level18 => 18,
            ZstdCompressionLevel::Level19 => 19,
            ZstdCompressionLevel::Level20 => 20,
            ZstdCompressionLevel::Level21 => 21,
            ZstdCompressionLevel::Max => 22,
        }
    }
}

#[cfg(feature = "parquet")]
impl From<ZstdCompressionLevel> for ::parquet::basic::Compression {
    fn from(value: ZstdCompressionLevel) -> Self {
        if let ZstdCompressionLevel::Uncompressed = value {
            Self::UNCOMPRESSED
        } else {
            let level = i32::from(u8::from(value));
            let level = ::parquet::basic::ZstdLevel::try_new(level)
                .expect("our range to mimic the parquet zstd range");
            Self::ZSTD(level)
        }
    }
}

#[derive(Debug, Clone)]
enum FragmentData {
    Json {
        builder: StateConfigBuilder,
    },
    #[cfg(feature = "parquet")]
    Parquet {
        tables: std::collections::HashMap<String, PathBuf>,
        compression: ZstdCompressionLevel,
    },
}
impl FragmentData {
    fn merge(mut self, data: FragmentData) -> anyhow::Result<Self> {
        match (&mut self, data) {
            (
                FragmentData::Json { builder, .. },
                FragmentData::Json {
                    builder: other_builder,
                    ..
                },
            ) => {
                builder.merge(other_builder);
            }
            #[cfg(feature = "parquet")]
            (
                FragmentData::Parquet {
                    tables,
                    compression,
                },
                FragmentData::Parquet {
                    tables: their_tables,
                    compression: their_compression,
                },
            ) => {
                tables.extend(their_tables);
                anyhow::ensure!(*compression == their_compression, "Fragments use different compressions.")
            }
            #[cfg(feature="parquet")]
            (a,b) => anyhow::bail!("Fragments don't have the same encoding and cannot be merged. Fragments: {a:?} and {b:?}"),
        };

        Ok(self)
    }
}

#[derive(Debug, Clone)]
pub struct SnapshotFragment {
    dir: PathBuf,
    data: FragmentData,
}

impl SnapshotFragment {
    pub fn merge(mut self, fragment: Self) -> anyhow::Result<Self> {
        self.data = self.data.merge(fragment.data)?;
        Ok(self)
    }

    pub fn finalize(
        self,
        latest_block_config: Option<LastBlockConfig>,
        chain_config: &ChainConfig,
    ) -> anyhow::Result<SnapshotMetadata> {
        let table_encoding = match self.data {
            FragmentData::Json { builder } => {
                let state_config = builder.build(latest_block_config)?;
                std::fs::create_dir_all(&self.dir)?;
                let state_file_path = self.dir.join("state_config.json");
                let file = std::fs::File::create(&state_file_path)?;
                serde_json::to_writer_pretty(file, &state_config)?;

                TableEncoding::Json {
                    filepath: state_file_path,
                }
            }
            #[cfg(feature = "parquet")]
            FragmentData::Parquet {
                tables,
                compression,
            } => {
                let latest_block_config_path =
                    self.dir.join("latest_block_config.parquet");
                SnapshotWriter::write_single_el_parquet(
                    &latest_block_config_path,
                    latest_block_config,
                    compression,
                )?;

                TableEncoding::Parquet {
                    tables,
                    latest_block_config_path,
                }
            }
        };

        SnapshotWriter::write_chain_config_and_metadata(
            &self.dir,
            chain_config,
            table_encoding,
        )
    }
}

impl SnapshotWriter {
    const CHAIN_CONFIG_FILENAME: &'static str = "chain_config.json";
    pub fn json(dir: impl Into<PathBuf>) -> Self {
        Self {
            encoder: EncoderType::Json {
                builder: StateConfigBuilder::default(),
            },
            dir: dir.into(),
        }
    }

    #[cfg(feature = "parquet")]
    pub fn parquet(
        dir: impl Into<::std::path::PathBuf>,
        compression_level: ZstdCompressionLevel,
    ) -> anyhow::Result<Self> {
        let dir = dir.into();
        std::fs::create_dir_all(&dir)?;
        Ok(Self {
            encoder: EncoderType::Parquet {
                table_encoders: TableEncoders::new(dir.clone(), compression_level),
                compression: compression_level,
            },
            dir,
        })
    }

    #[cfg(feature = "test-helpers")]
    pub fn write_state_config(
        mut self,
        state_config: crate::StateConfig,
        chain_config: &ChainConfig,
    ) -> anyhow::Result<SnapshotMetadata> {
        use fuel_core_storage::tables::{
            Coins,
            ContractsAssets,
            ContractsLatestUtxo,
            ContractsRawCode,
            ContractsState,
            Messages,
        };
        use fuel_core_types::fuel_vm::BlobData;

        use crate::AsTable;

        self.write::<Coins>(state_config.as_table())?;
        self.write::<Messages>(state_config.as_table())?;
        self.write::<BlobData>(state_config.as_table())?;
        self.write::<ContractsRawCode>(state_config.as_table())?;
        self.write::<ContractsLatestUtxo>(state_config.as_table())?;
        self.write::<ContractsState>(state_config.as_table())?;
        self.write::<ContractsAssets>(state_config.as_table())?;
        self.close(state_config.last_block, chain_config)
    }

    pub fn write<T>(&mut self, elements: Vec<TableEntry<T>>) -> anyhow::Result<()>
    where
        T: TableWithBlueprint,
        TableEntry<T>: serde::Serialize,
        StateConfigBuilder: AddTable<T>,
    {
        match &mut self.encoder {
            EncoderType::Json { builder, .. } => {
                builder.add(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet { table_encoders, .. } => {
                table_encoders.encoder::<T>()?.write::<T>(elements)
            }
        }
    }

    #[cfg(feature = "parquet")]
    fn write_single_el_parquet(
        path: &std::path::Path,
        data: impl serde::Serialize,
        compression: ZstdCompressionLevel,
    ) -> anyhow::Result<()> {
        let mut encoder = parquet::encode::Encoder::new(
            std::fs::File::create(path)?,
            compression.into(),
        )?;
        encoder.write(vec![postcard::to_stdvec(&data)?])?;
        encoder.close()
    }

    pub fn close(
        self,
        latest_block_config: Option<LastBlockConfig>,
        chain_config: &ChainConfig,
    ) -> anyhow::Result<SnapshotMetadata> {
        self.partial_close()?
            .finalize(latest_block_config, chain_config)
    }

    fn write_chain_config_and_metadata(
        dir: &std::path::Path,
        chain_config: &ChainConfig,
        table_encoding: TableEncoding,
    ) -> anyhow::Result<SnapshotMetadata> {
        let chain_config_path = dir.join(Self::CHAIN_CONFIG_FILENAME);
        chain_config.write(&chain_config_path)?;

        let metadata = SnapshotMetadata {
            chain_config: chain_config_path,
            table_encoding,
        };
        metadata.clone().write(dir)?;
        Ok(metadata)
    }

    pub fn partial_close(self) -> anyhow::Result<SnapshotFragment> {
        let data = match self.encoder {
            EncoderType::Json { builder } => FragmentData::Json { builder },
            #[cfg(feature = "parquet")]
            EncoderType::Parquet {
                table_encoders,
                compression,
                ..
            } => {
                let tables = table_encoders.close()?;
                FragmentData::Parquet {
                    tables,
                    compression,
                }
            }
        };
        let snapshot_fragment = SnapshotFragment {
            dir: self.dir,
            data,
        };
        Ok(snapshot_fragment)
    }
}

#[cfg(feature = "parquet")]
struct PostcardParquetEncoder {
    path: PathBuf,
    encoder: parquet::encode::Encoder<std::fs::File>,
}

#[cfg(feature = "parquet")]
impl PostcardParquetEncoder {
    pub fn new(path: PathBuf, encoder: parquet::encode::Encoder<std::fs::File>) -> Self {
        Self { path, encoder }
    }

    fn write<T>(&mut self, elements: Vec<TableEntry<T>>) -> anyhow::Result<()>
    where
        T: fuel_core_storage::Mappable,
        TableEntry<T>: serde::Serialize,
    {
        use itertools::Itertools;
        let encoded: Vec<_> = elements
            .into_iter()
            .map(|entry| postcard::to_stdvec(&entry))
            .try_collect()?;
        self.encoder.write(encoded)
    }
}

#[cfg(feature = "parquet")]
struct TableEncoders {
    dir: PathBuf,
    compression: ZstdCompressionLevel,
    encoders: std::collections::HashMap<String, PostcardParquetEncoder>,
}

#[cfg(feature = "parquet")]
impl TableEncoders {
    fn new(dir: PathBuf, compression: ZstdCompressionLevel) -> Self {
        Self {
            dir,
            compression,
            encoders: Default::default(),
        }
    }

    fn encoder<T: fuel_core_storage::structured_storage::TableWithBlueprint>(
        &mut self,
    ) -> anyhow::Result<&mut PostcardParquetEncoder> {
        use fuel_core_storage::kv_store::StorageColumn;

        let name = StorageColumn::name(&T::column()).to_string();

        let encoder = match self.encoders.entry(name) {
            std::collections::hash_map::Entry::Occupied(encoder) => encoder.into_mut(),
            std::collections::hash_map::Entry::Vacant(vacant) => {
                let name = vacant.key();
                let file_path = self.dir.join(format!("{name}.parquet"));
                let file = std::fs::File::create(&file_path)?;
                let encoder = PostcardParquetEncoder::new(
                    file_path,
                    parquet::encode::Encoder::new(file, self.compression.into())?,
                );
                vacant.insert(encoder)
            }
        };

        Ok(encoder)
    }

    fn close(self) -> anyhow::Result<std::collections::HashMap<String, PathBuf>> {
        let mut files = std::collections::HashMap::new();
        for (file, encoder) in self.encoders {
            encoder.encoder.close()?;
            files.insert(file, encoder.path);
        }
        Ok(files)
    }
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use fuel_core_storage::{
        kv_store::StorageColumn,
        structured_storage::TableWithBlueprint,
        tables::{
            Coins,
            ContractsAssets,
            ContractsLatestUtxo,
            ContractsRawCode,
            ContractsState,
            Messages,
        },
    };
    use rand::{
        rngs::StdRng,
        SeedableRng,
    };

    use crate::StateConfig;

    use super::*;

    #[test]
    fn can_roundtrip_compression_level() {
        use strum::IntoEnumIterator;

        for level in crate::ZstdCompressionLevel::iter() {
            let u8_level = u8::from(level);
            let roundtrip = ZstdCompressionLevel::try_from(u8_level).unwrap();
            assert_eq!(level, roundtrip);
        }
    }

    #[test]
    fn parquet_encoder_encodes_tables_in_expected_files() {
        fn file_created_and_present_in_metadata<T>()
        where
            T: TableWithBlueprint,
            T::OwnedKey: serde::Serialize,
            T::OwnedValue: serde::Serialize,
            StateConfigBuilder: AddTable<T>,
        {
            // given
            use pretty_assertions::assert_eq;

            let dir = tempfile::tempdir().unwrap();
            let mut writer =
                SnapshotWriter::parquet(dir.path(), ZstdCompressionLevel::Uncompressed)
                    .unwrap();

            // when
            writer.write::<T>(vec![]).unwrap();
            let snapshot = writer.close(None, &ChainConfig::local_testnet()).unwrap();

            // then
            assert!(snapshot.chain_config.exists());
            let TableEncoding::Parquet { tables, .. } = snapshot.table_encoding else {
                panic!("Expected parquet encoding")
            };
            assert_eq!(tables.len(), 1, "Expected single table");
            let (table_name, path) = tables.into_iter().next().unwrap();

            assert_eq!(table_name, T::column().name());
            assert!(dir.path().join(path).exists());
        }

        file_created_and_present_in_metadata::<Coins>();
        file_created_and_present_in_metadata::<Messages>();
        file_created_and_present_in_metadata::<ContractsRawCode>();
        file_created_and_present_in_metadata::<ContractsLatestUtxo>();
        file_created_and_present_in_metadata::<ContractsState>();
        file_created_and_present_in_metadata::<ContractsAssets>();
    }

    #[test]
    fn all_compressions_are_valid() {
        use ::parquet::basic::Compression;
        use strum::IntoEnumIterator;
        for level in ZstdCompressionLevel::iter() {
            let _ = Compression::from(level);
        }
    }

    #[test]
    fn json_snapshot_is_human_readable() {
        // given
        use crate::Randomize;
        let dir = tempfile::tempdir().unwrap();
        let writer = SnapshotWriter::json(dir.path());
        let mut rng = StdRng::from_seed([0; 32]);
        let state = StateConfig::randomize(&mut rng);

        // when
        let snapshot = writer
            .write_state_config(state, &ChainConfig::local_testnet())
            .unwrap();

        // then
        let TableEncoding::Json { filepath } = snapshot.table_encoding else {
            panic!("Expected json encoding")
        };
        let encoded_json = std::fs::read_to_string(filepath).unwrap();

        insta::assert_snapshot!(encoded_json);
    }

    fn given_parquet_writer(path: &Path) -> SnapshotWriter {
        SnapshotWriter::parquet(path, ZstdCompressionLevel::Uncompressed).unwrap()
    }

    fn given_json_writer(path: &Path) -> SnapshotWriter {
        SnapshotWriter::json(path)
    }

    #[test_case::test_case(given_parquet_writer)]
    #[test_case::test_case(given_json_writer)]
    fn can_partially_close_without_chain_and_block_height(
        writer: impl Fn(&Path) -> SnapshotWriter + Copy,
    ) {
        // given
        let dir = tempfile::tempdir().unwrap();
        let writer = writer(dir.path());

        // when
        let result = writer.partial_close();

        // then
        assert!(result.is_ok());
    }

    #[test]
    fn merging_json_and_parquet_fragments_fails() {
        // given
        let dir = tempfile::tempdir().unwrap();
        let json_writer = super::SnapshotWriter::json(dir.path());
        let parquet_writer = super::SnapshotWriter::parquet(
            dir.path(),
            super::ZstdCompressionLevel::Uncompressed,
        );

        let json_fragment = json_writer.partial_close().unwrap();
        let parquet_fragment = parquet_writer.unwrap().partial_close().unwrap();

        {
            // when
            let result = json_fragment.clone().merge(parquet_fragment.clone());

            // when
            let err = result.unwrap_err();
            assert!(err.to_string().contains(
                "Fragments don't have the same encoding and cannot be merged."
            ));
        }
        {
            // when
            let result = parquet_fragment.merge(json_fragment);

            // when
            let err = result.unwrap_err();
            assert!(err.to_string().contains(
                "Fragments don't have the same encoding and cannot be merged."
            ));
        }
    }

    // It is enough just for the test to compile
    #[test]
    #[ignore]
    fn fragment_must_be_send_sync() {
        fn _assert_send<T: Send + Sync>() {}
        _assert_send::<super::SnapshotFragment>();
    }
}
