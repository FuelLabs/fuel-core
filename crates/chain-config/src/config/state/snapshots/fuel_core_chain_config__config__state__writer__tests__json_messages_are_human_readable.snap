---
source: crates/chain-config/src/config/state/writer.rs
expression: encoded_json
---
{
  "coins": [],
  "messages": [
    {
      "sender": "0101010101010101010101010101010101010101010101010101010101010101",
      "recipient": "0202020202020202020202020202020202020202020202020202020202020202",
      "nonce": "0303030303030303030303030303030303030303030303030303030303030303",
      "amount": 4,
      "data": "0505050505050505050505050505050505050505050505050505050505050505",
      "da_height": 6
    }
  ],
  "contracts": [],
  "contract_state": [],
  "contract_balance": [],
  "block_height": 0
}
